src/frontend/src/expr/window_function.rs

```diff
@@ -29,86 +27,104 @@ use super::{Expr, ExprImpl, OrderBy, RwResult};
 #[derive(Clone, Eq, PartialEq, Hash)]
 pub struct WindowFunction {
+    pub kind: WindowFuncKind,
     pub args: Vec<ExprImpl>,
     pub return_type: DataType,
-    pub function_type: WindowFunctionType,
     pub partition_by: Vec<ExprImpl>,
     pub order_by: OrderBy,
-}
-
-#[derive(Debug, Display, Copy, Clone, PartialEq, Eq, Hash)]
-#[display(style = "SNAKE_CASE")]
-pub enum WindowFunctionType {
-    RowNumber,
-    Rank,
-    DenseRank,
-}
-
-impl WindowFunctionType {
-    pub fn is_rank_function(&self) -> bool {
-        matches!(
-            self,
-            WindowFunctionType::RowNumber
-                | WindowFunctionType::Rank
-                | WindowFunctionType::DenseRank
-        )
-    }
-}
-
-impl FromStr for WindowFunctionType {
-    type Err = ErrorCode;
-
-    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
-        match s.to_ascii_lowercase().as_str() {
-            "row_number" => Ok(WindowFunctionType::RowNumber),
-            "rank" => Ok(WindowFunctionType::Rank),
-            "dense_rank" => Ok(WindowFunctionType::DenseRank),
-            _ => Err(ErrorCode::NotImplemented(
-                format!("unknown table function kind: {s}"),
-                None.into(),
-            )),
-        }
-    }
+    pub frame: Option<Frame>,
 }
 
 impl WindowFunction {
     /// Create a `WindowFunction` expr with the return type inferred from `func_type` and types of
     /// `inputs`.
     pub fn new(
-        function_type: WindowFunctionType,
+        kind: WindowFuncKind,
         partition_by: Vec<ExprImpl>,
         order_by: OrderBy,
         args: Vec<ExprImpl>,
+        frame: Option<Frame>,
     ) -> RwResult<Self> {
-        if !args.is_empty() {
-            return Err(ErrorCode::BindError(format!(
-                "the length of args of {function_type} function should be 0"
-            ))
-            .into());
-        }
-
+        let return_type = Self::infer_return_type(kind, &args)?;
         Ok(Self {
+            kind,
             args,
-            return_type: DataType::Int64,
-            function_type,
+            return_type,
             partition_by,
             order_by,
+            frame,
         })
     }
+
+    fn infer_return_type(kind: WindowFuncKind, args: &[ExprImpl]) -> RwResult<DataType> {
+        use WindowFuncKind::*;
+        match (kind, args) {
+            (RowNumber, []) => Ok(DataType::Int64),
+            (Rank, []) => Ok(DataType::Int64),
+            (DenseRank, []) => Ok(DataType::Int64),
+
+            (Lag | Lead, [value]) => Ok(value.return_type()),
+            (Lag | Lead, [value, offset]) => {
+                if !offset.return_type().is_int() {
+                    return Err(ErrorCode::InvalidInputSyntax(format!(
+                        "the `offset` of `{kind}` function should be integer"
+                    ))
+                    .into());
+                }
+                if offset.as_literal().is_none() {
```

Comment:
As we evaluate it with `eval_row_const` later, how about determining this by `is_const` as well?